pymongo/operations.py

```diff
@@ -80,48 +81,121 @@ class _Op(str, enum.Enum):
 class InsertOne(Generic[_DocumentType]):
     """Represents an insert_one operation."""
 
-    __slots__ = ("_doc",)
+    __slots__ = (
+        "_doc",
+        "_namespace",
+    )
 
-    def __init__(self, document: _DocumentType) -> None:
+    def __init__(self, document: _DocumentType, namespace: Optional[str] = None) -> None:
         """Create an InsertOne instance.
 
-        For use with :meth:`~pymongo.asynchronous.collection.AsyncCollection.bulk_write` and :meth:`~pymongo.collection.Collection.bulk_write`.
+        For use with :meth:`~pymongo.asynchronous.collection.AsyncCollection.bulk_write`, :meth:`~pymongo.collection.Collection.bulk_write`,
+        :meth:`~pymongo.asynchronous.mongo_client.AsyncMongoClient.bulk_write` and :meth:`~pymongo.mongo_client.MongoClient.bulk_write`.
 
         :param document: The document to insert. If the document is missing an
             _id field one will be added.
+        :param namespace: (optional) The namespace in which to insert a document.
+
+        .. versionchanged:: 4.9
+           Added the `namespace` option to support `MongoClient.bulk_write`.
         """
         self._doc = document
+        self._namespace = namespace
 
     def _add_to_bulk(self, bulkobj: _AgnosticBulk) -> None:
         """Add this operation to the _AsyncBulk/_Bulk instance `bulkobj`."""
         bulkobj.add_insert(self._doc)  # type: ignore[arg-type]
 
+    def _add_to_client_bulk(self, bulkobj: _AgnosticClientBulk) -> None:
+        """Add this operation to the _AsyncClientBulk/_ClientBulk instance `bulkobj`."""
+        bulkobj.add_insert(self._namespace, self._doc)  # type: ignore[arg-type]
```

Comment:
It should be an error to call client.bulk_write() with an operation that doesn't have a namespace. We should add a check and remove the type ignore here. Could you add a test case for this?
```python 
client.bulk_write([InsertOne({})]) # Should raise error due to missing namespace.
```

Reply:
Also related, does the spec say to perform any validation on the namespace?

Reply:
The spec says to ensure that the namespace is provided for each write request, but apart from that it doesn't mention any additional validation to perform